File: js/src/builtin/Module.cpp
/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 4 -*-
 * vim: set ts=8 sts=4 et sw=4 tw=99:
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "jsobjinlines.h"
#include "builtin/Module.h"

using namespace js;

Class js::ModuleClass = {
    "Module",
    JSCLASS_HAS_RESERVED_SLOTS(2) | JSCLASS_IS_ANONYMOUS,
    JS_PropertyStub,        /* addProperty */
    JS_DeletePropertyStub,  /* delProperty */
    JS_PropertyStub,        /* getProperty */
    JS_StrictPropertyStub,  /* setProperty */
    JS_EnumerateStub,
    JS_ResolveStub,
    JS_ConvertStub
};

inline void
Module::setAtom(JSAtom *atom)
{
    setReservedSlot(ATOM_SLOT, StringValue(atom));
}

inline void
Module::setScript(JSScript *script)
{
    setReservedSlot(SCRIPT_SLOT, PrivateValue(script));
}

Module *
Module::create(JSContext *cx, HandleAtom atom)
{
    RootedObject object(cx, NewBuiltinClassInstance(cx, &ModuleClass));
    if (!object)
        return NULL;
    RootedModule module(cx, &object->asModule());
    module->setAtom(atom);
    module->setScript(NULL);
    return module;
}

File: js/src/builtin/Module.js
/* -*- Mode: javascript; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// https://tc39.es/proposal-top-level-await/#sec-gather-async-parent-completions
function GatherAsyncParentCompletions(module, execList = new_List()) {
  assert(module.status == MODULE_STATUS_EVALUATED, "bad status for async module");

  // Step 5.
  // asyncParentModules is a list, and doesn't have a .length. Might be worth changing
  // later on.
  let i = 0;
  while (module.asyncParentModules[i]) {
    const m = module.asyncParentModules[i];
    if (GetCycleRoot(m).status != MODULE_STATUS_EVALUATED_ERROR &&
        !callFunction(std_Array_includes, execList, m)) {
      assert(!m.evaluationError, "should not have evaluation error");
      assert(m.pendingAsyncDependencies > 0, "should have at least one dependency");
      UnsafeSetReservedSlot(m,
                            MODULE_OBJECT_PENDING_ASYNC_DEPENDENCIES_SLOT,
                            m.pendingAsyncDependencies - 1);
      if (m.pendingAsyncDependencies === 0) {
        callFunction(std_Array_push, execList, m);
        if (!m.async) {
          execList = GatherAsyncParentCompletions(m, execList);
        }
      }
    }
    i++;
  }
  callFunction(ArraySort,
               execList,
               (a, b) => a.asyncEvaluatingPostOrder - b.asyncEvaluatingPostOrder);
  return execList
}
